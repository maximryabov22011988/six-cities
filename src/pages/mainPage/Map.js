import React, { Component } from 'react';
import PropTypes from 'prop-types';
import leaflet from 'leaflet';

const mapStyles = {
  top: '15px',
  height: '794px'
};

const DEFAULT_MAP_ZOOM = 12;

const LEAFLET_LAYER = {
  URL_TEMPLATE:
    'https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png',
  OPTIONS: {
    attribution:
      '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>'
  }
};

const propTypes = {
  offers: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      image: PropTypes.string,
      price: PropTypes.number,
      type: PropTypes.oneOf(['Apartment', 'Private room']),
      rating: PropTypes.number,
      isPremium: PropTypes.bool,
      isBookmark: PropTypes.bool,
      location: PropTypes.arrayOf(PropTypes.number)
    })
  ).isRequired,
  city: PropTypes.arrayOf(PropTypes.number).isRequired
};

class Map extends Component {
  constructor(props) {
    super(props);

    this.settings = {
      center: props.city,
      zoom: DEFAULT_MAP_ZOOM,
      zoomControl: false,
      marker: true
    };
  }

  componentDidMount() {
    this.initCard();
  }

  get pinIcon() {
    return leaflet.icon({
      iconUrl: 'img/pin.svg',
      iconSize: [30, 40]
    });
  }

  addPins() {
    const { offers } = this.props;
    const icon = this.pinIcon;
    offers.forEach(({ location, title }) => {
      leaflet.marker(location, { icon, title }).addTo(this.offerMap);
    });
  }

  initCard() {
    const { settings } = this;
    const { city } = this.props;

    this.offerMap = leaflet.map('map', settings);
    this.offerMap.setView(city, DEFAULT_MAP_ZOOM);
    leaflet
      .tileLayer(LEAFLET_LAYER.URL_TEMPLATE, LEAFLET_LAYER.OPTIONS)
      .addTo(this.offerMap);

    this.addPins();
  }

  render() {
    return <div id="map" style={mapStyles} />;
  }
}

Map.propTypes = propTypes;

export default Map;
