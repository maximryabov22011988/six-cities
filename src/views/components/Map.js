import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Leaflet from 'leaflet';

const radiusInKm = 3;

const circleStyles = {
  color: '#FF9000',
  fillColor: '#FF9000',
  fillOpacity: 0.1,
  radius: radiusInKm * 1000,
};

const mapStyles = {
  fixed: {
    top: 0,
    zIndex: 100,
    width: '1144px',
    height: '580px',
    marginLeft: 'auto',
    marginRight: 'auto',
  },
  flexible: {
    top: 0,
    height: '100%',
    zIndex: 100,
  },
};

const mapContainerId = 'map';

const LEAFLET_LAYER = {
  URL_TEMPLATE: 'https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png',
  OPTIONS: {
    attribution:
      '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>',
  },
};

const propTypes = {
  className: PropTypes.string,
  currentOffer: PropTypes.shape({
    bedrooms: PropTypes.number,
    city: PropTypes.shape({
      name: PropTypes.string,
      location: PropTypes.object,
    }),
    description: PropTypes.string,
    goods: PropTypes.arrayOf(PropTypes.string),
    host: PropTypes.shape({
      avatar_url: PropTypes.string,
      id: PropTypes.number,
      is_pro: PropTypes.bool,
      name: PropTypes.string,
    }),
    id: PropTypes.number,
    images: PropTypes.arrayOf(PropTypes.string),
    is_favorite: PropTypes.bool,
    is_premium: PropTypes.bool,
    location: PropTypes.shape({
      latitude: PropTypes.number,
      longitude: PropTypes.number,
      zoom: PropTypes.number,
    }),
    max_adults: PropTypes.number,
    preview_image: PropTypes.string,
    price: PropTypes.number,
    rating: PropTypes.number,
    title: PropTypes.string,
    type: PropTypes.string,
  }),
  offers: PropTypes.arrayOf(
    PropTypes.shape({
      bedrooms: PropTypes.number,
      city: PropTypes.shape({
        name: PropTypes.string,
        location: PropTypes.object,
      }),
      description: PropTypes.string,
      goods: PropTypes.arrayOf(PropTypes.string),
      host: PropTypes.shape({
        avatar_url: PropTypes.string,
        id: PropTypes.number,
        is_pro: PropTypes.bool,
        name: PropTypes.string,
      }),
      id: PropTypes.number,
      images: PropTypes.arrayOf(PropTypes.string),
      is_favorite: PropTypes.bool,
      is_premium: PropTypes.bool,
      location: PropTypes.shape({
        latitude: PropTypes.number,
        longitude: PropTypes.number,
        zoom: PropTypes.number,
      }),
      max_adults: PropTypes.number,
      preview_image: PropTypes.string,
      price: PropTypes.number,
      rating: PropTypes.number,
      title: PropTypes.string,
      type: PropTypes.string,
    }),
  ),
};

class Map extends Component {
  componentDidMount() {
    const { currentOffer, offers } = this.props;

    const { latitude, longitude, zoom } = offers ? offers[0].city.location : currentOffer.location;

    const settings = {
      center: [latitude, longitude],
      zoom,
      zoomControl: false,
    };

    this.offerMap = Leaflet.map(mapContainerId, settings);
    this.offerMap.setView([latitude, longitude], zoom);
    Leaflet.tileLayer(LEAFLET_LAYER.URL_TEMPLATE, LEAFLET_LAYER.OPTIONS).addTo(this.offerMap);
    this.pinGroup = Leaflet.layerGroup().addTo(this.offerMap);
    this.addPins(this.pinGroup);
  }

  componentDidUpdate() {
    const { currentOffer, offers } = this.props;

    const { latitude, longitude, zoom } = offers ? offers[0].city.location : currentOffer.location;

    if (this.offerMap) {
      this.offerMap.setView([latitude, longitude], zoom);
      this.pinGroup.clearLayers();
      this.addPins(this.pinGroup);
    }
  }

  get pinIcon() {
    return Leaflet.icon({
      iconUrl: 'img/pin.svg',
      iconSize: [27, 39],
      iconAnchor: [13, 39],
    });
  }

  get activePinIcon() {
    return Leaflet.icon({
      iconUrl: 'img/pin-active.svg',
      iconSize: [27, 39],
      iconAnchor: [13, 39],
    });
  }

  addPins(group) {
    const { currentOffer, offers } = this.props;
    const icon = this.pinIcon;
    const activeIcon = this.activePinIcon;

    if (currentOffer) {
      const {
        location: { latitude, longitude },
      } = currentOffer;
      Leaflet.marker([latitude, longitude], { icon: activeIcon }).addTo(group);
      Leaflet.circle([latitude, longitude], circleStyles).addTo(group);
    }

    offers.forEach(({ location: { latitude, longitude } }) => {
      Leaflet.marker([latitude, longitude], { icon }).addTo(group);
    });
  }

  render() {
    const { className, fixed } = this.props;
    return <div className={className} id={mapContainerId} style={fixed ? mapStyles.fixed : mapStyles.flexible} />;
  }
}

Map.propTypes = propTypes;

export default Map;
