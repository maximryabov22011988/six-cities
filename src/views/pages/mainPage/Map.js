import React, { Component } from 'react';
// import PropTypes from 'prop-types';
import Leaflet from 'leaflet';

const mapStyles = {
  top: '15px',
  height: '794px'
};

const LEAFLET_LAYER = {
  URL_TEMPLATE:
    'https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png',
  OPTIONS: {
    attribution:
      '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>'
  }
};

// const propTypes = {
//   offers: PropTypes.arrayOf(
//     PropTypes.shape({
//       id: PropTypes.string,
//       title: PropTypes.string,
//       image: PropTypes.string,
//       price: PropTypes.number,
//       type: PropTypes.oneOf(['Apartment', 'Private room']),
//       rating: PropTypes.number,
//       isPremium: PropTypes.bool,
//       isBookmark: PropTypes.bool,
//       location: PropTypes.arrayOf(PropTypes.number)
//     })
//   ).isRequired
// };

class Map extends Component {
  constructor(props) {
    super(props);

    this.leaflet = Leaflet;

    this.icon = this.leaflet.icon({
      iconUrl: 'img/pin.svg',
      iconSize: [27, 39],
      iconAnchor: [13, 39]
    });
  }

  componentDidMount() {
    const { offers } = this.props;

    if (offers.length > 0) {
      this.map = this.leaflet.map(
        'map',
        {
          center: [
            offers[0].city.location.latitude,
            offers[0].city.location.longitude
          ],
          zoom: offers[0].city.location.zoom,
          zoomControl: false,
          layers: new this.leaflet.TileLayer(
            LEAFLET_LAYER.URL_TEMPLATE,
            LEAFLET_LAYER.OPTIONS
          )
        },
        100
      );

      this.markers = this.leaflet.layerGroup().addTo(this.map);
      this.addMarkers(offers, this.markers);
    }
  }

  componentDidUpdate() {
    const { offers } = this.props;

    if (this.map) {
      this.map.setView(
        [offers[0].city.location.latitude, offers[0].city.location.longitude],
        offers[0].city.location.zoom
      );
      this.markers.clearLayers();
      this.addMarkers(offers, this.markers);
    }
  }

  addMarkers(offers, group) {
    offers.map(offer => {
      this.leaflet
        .marker([offer.location.latitude, offer.location.longitude], {
          icon: this.icon
        })
        .addTo(group);
    });
  }

  render() {
    return <div id="map" style={mapStyles} />;
  }
}

// Map.propTypes = propTypes;

export default Map;
