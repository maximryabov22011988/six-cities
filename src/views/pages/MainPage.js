import React from 'react';
import PropTypes from 'prop-types';

import Page from '../components/Page';
import Header from '../components/Header';
import Logo from '../components/Logo';
import UserInfo from '../components/UserInfo';

import Content from './mainPage/Content';
import Nav from './mainPage/Nav';
import NavList from './mainPage/NavList';
import PlacesContainer from './mainPage/PlacesContainer';
import Places from './mainPage/Places';
import EmptyPlaces from './mainPage/EmptyPlaces';
import Map from '../components/Map';

const propTypes = {
  changeCity: PropTypes.func.isRequired,
  changeSorting: PropTypes.func.isRequired,
  cities: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      location: PropTypes.arrayOf(PropTypes.number),
      zoom: PropTypes.number,
    })
  ).isRequired,
  currentCity: PropTypes.shape({
    name: PropTypes.string,
    location: PropTypes.arrayOf(PropTypes.number),
  }).isRequired,
  isAuthUser: PropTypes.bool.isRequired,
  offers: PropTypes.arrayOf(PropTypes.object).isRequired,
  user: PropTypes.shape({
    id: PropTypes.number,
    email: PropTypes.string,
    name: PropTypes.string,
    avatar_url: PropTypes.string,
    is_pro: PropTypes.bool,
  }).isRequired,
};

class MainPage extends React.Component {
  state = {
    activeOffer: null,
  };

  getSearchResultsText() {
    const { currentCity, offers } = this.props;
    return `${offers.length} places to stay in ${currentCity.name}`;
  }

  handleActiveOfferClick = (id) => (evt) => {
    evt.preventDefault();
    this.setState({
      activeOffer: id,
    });
  };

  handleChangeCity = (city) => (evt) => {
    evt.preventDefault();
    const { changeCity } = this.props;
    changeCity(city);
  };

  handleChangeSorting = (sorting) => {
    const { changeSorting } = this.props;
    changeSorting(sorting);
  };

  havePlaceCards() {
    const { offers } = this.props;
    return offers.length > 0;
  }

  render() {
    const { user, isAuthUser, currentCity, cities, offers } = this.props;
    const { activeOffer } = this.state;

    return (
      <Page parentClassName="main">
        <Header
          logo={<Logo isActive position="header" />}
          userInfo={<UserInfo avatarUrl={user.avatar_url} email={user.email} isAuth={isAuthUser} />}
        />
        <Content isEmpty={!this.havePlaceCards()}>
          <Nav>
            <NavList cities={cities} currentCity={currentCity.name} onChangeCity={this.handleChangeCity} />
          </Nav>
          <PlacesContainer
            isEmpty={!this.havePlaceCards()}
            leftPanel={
              this.havePlaceCards() ? (
                <Places
                  offers={offers}
                  searchResultText={this.getSearchResultsText()}
                  onActiveOfferClick={this.handleActiveOfferClick}
                  onChangeSorting={this.handleChangeSorting}
                />
              ) : (
                <EmptyPlaces city={currentCity.name} />
              )
            }
            rightPanel={<Map activeOffer={activeOffer} offers={offers} />}
          />
        </Content>
      </Page>
    );
  }
}

MainPage.propTypes = propTypes;

export default MainPage;
