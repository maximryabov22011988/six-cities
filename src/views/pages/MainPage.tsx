import * as React from 'react';

import Page from '../components/Page';
import Header from '../components/Header';
import Logo from '../components/Logo';
import UserInfo from '../components/UserInfo';

import Content from './mainPage/Content';
import Nav from './mainPage/Nav';
import NavList from './mainPage/NavList';
import PlacesContainer from './mainPage/PlacesContainer';
import Places from './mainPage/Places';
import EmptyPlaces from './mainPage/EmptyPlaces';
import Map from '../components/Map';

import { OfferCity, City, User, Offer } from '../interfaces';

interface Props {
  addOfferToFavorities: (hotelId: number) => void;
  changeCity: (city: City) => void;
  changeSorting: (sorting: string) => void;
  cities: Array<City>;
  currentCity: OfferCity;
  isAuthUser: boolean;
  offers: Array<Offer>;
  removeOfferFromFavorities: (hotelId: number) => void;
  user: User;
}

interface State {
  activeOffer: number;
}

class MainPage extends React.Component<Props, State> {
  state = {
    activeOffer: null,
  };

  getSearchResultsText() {
    const { currentCity, offers } = this.props;
    return `${offers.length} places to stay in ${currentCity.name}`;
  }

  handleActiveOfferClick = (id) => (evt) => {
    evt.preventDefault();
    this.setState({
      activeOffer: id,
    });
  };

  handleAddToFavorities = (hotelId) => () => {
    const { addOfferToFavorities } = this.props;
    addOfferToFavorities(hotelId);
  };

  handleRemoveFromFavorities = (hotelId) => () => {
    const { removeOfferFromFavorities } = this.props;
    removeOfferFromFavorities(hotelId);
  };

  handleChangeCity = (city) => (evt) => {
    evt.preventDefault();
    const { changeCity } = this.props;
    changeCity(city);
  };

  handleChangeSorting = (sorting) => {
    const { changeSorting } = this.props;
    changeSorting(sorting);
  };

  havePlaceCards() {
    const { offers } = this.props;
    return offers.length > 0;
  }

  render() {
    const { user, isAuthUser, currentCity, cities, offers } = this.props;
    const { activeOffer } = this.state;

    return (
      <Page parentClassName="main">
        <Header
          logo={<Logo isActive position="header" />}
          userInfo={<UserInfo avatarUrl={user.avatar_url} email={user.email} isAuth={isAuthUser} />}
        />
        <Content isEmpty={!this.havePlaceCards()}>
          <Nav>
            <NavList cities={cities} currentCity={currentCity.name} onChangeCity={this.handleChangeCity} />
          </Nav>
          <PlacesContainer
            isEmpty={!this.havePlaceCards()}
            leftPanel={
              this.havePlaceCards() ? (
                <Places
                  offers={offers}
                  searchResultText={this.getSearchResultsText()}
                  onActiveOfferClick={this.handleActiveOfferClick}
                  onAddToFavorities={this.handleAddToFavorities}
                  onChangeSorting={this.handleChangeSorting}
                  onRemoveFromFavorities={this.handleRemoveFromFavorities}
                />
              ) : (
                <EmptyPlaces city={currentCity.name} />
              )
            }
            rightPanel={<Map activeOffer={activeOffer} offers={offers} />}
          />
        </Content>
      </Page>
    );
  }
}

export default MainPage;
