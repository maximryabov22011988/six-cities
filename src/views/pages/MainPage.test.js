import React from 'react';
import renderer from 'react-test-renderer';

import MainPage from './MainPage';
import Map from './mainPage/Map';

const userMock = {
  avatarUrl: '/static/avatar/1.jpg',
  email: 'Oliver.conner@gmail.com',
};

const currentCityMock = {
  name: 'Cologne',
  location: [50.938361, 6.959974],
};

const citiesMock = [
  {
    name: 'Cologne',
    location: [50.938361, 6.959974],
    zoom: 13,
  },
];

const offersMock = [
  {
    id: 1,
    city: {
      name: 'Cologne',
      location: {
        latitude: 52.37454,
        longitude: 4.897976,
        zoom: 13,
      },
    },
    title: 'The Pondhouse - A Magical Place',
    preview_image: `img/apartment-01.jpg`,
    price: 120,
    type: 'house',
    rating: 4.6,
    is_premium: true,
    is_favorite: false,
  },
];

it('MainPage correctly render', () => {
  Map.prototype.componentDidMount = jest.fn();
  const mainPage = renderer
    .create(
      <MainPage
        user={userMock}
        currentCity={currentCityMock}
        cities={citiesMock}
        offers={offersMock}
        changeCity={jest.fn()}
        isAuthUser
      />,
    )
    .toJSON();

  expect(mainPage).toMatchSnapshot();
});
