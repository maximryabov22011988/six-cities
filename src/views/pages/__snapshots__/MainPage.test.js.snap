// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage correctly render 1`] = `
<div
  className="page page--gray page--main"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link header__logo-link--active"
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <div
                className="header__nav-link header__nav-link--profile"
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                >
                  <img
                    alt="avatar"
                    className="user__avatar"
                    src="https://es31-server.appspot.com/six-citiesundefined"
                  />
                </div>
                <span
                  className="header__user-name user__name"
                >
                  Oliver.conner@gmail.com
                </span>
              </div>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--index"
  >
    <h1
      className="visually-hidden"
    >
      Cities
    </h1>
    <div
      className="cities tabs"
    >
      <section
        className="locations container"
      >
        <ul
          className="locations__list tabs__list"
        >
          <li
            className="locations__item"
          >
            <a
              className="tabs__item tabs__item--active locations__item-link"
              href="#"
              onClick={[Function]}
            >
              <span>
                Cologne
              </span>
            </a>
          </li>
        </ul>
      </section>
    </div>
    <div
      className="cities__places-wrapper"
    >
      <div
        className="cities__places-container container"
      >
        <section
          className="cities__places places"
        >
          <h2
            className="visually-hidden"
          >
            Places
          </h2>
          <b
            className="places__found"
          >
            1 places to stay in Cologne
          </b>
          <div
            className="places__sorting"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex="0"
          >
            <span
              className="places__sorting-caption"
            >
              Sort by
            </span>
            <span
              className="places__sorting-type"
              onClick={[Function]}
              onFocus={[Function]}
              tabIndex="0"
            >
              Popular
              <svg
                className="places__sorting-arrow"
                height="4"
                width="7"
              >
                <use
                  xlinkHref="#icon-arrow-select"
                />
              </svg>
            </span>
          </div>
          <div
            className="cities__places-list places__list tabs__content"
          >
            <article
              className="cities__place-card place-card"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__mark"
              >
                <span>
                  Premium
                </span>
              </div>
              <div
                className="cities__image-wrapper place-card__image-wrapper"
              >
                <a
                  href="#"
                  onClick={[Function]}
                >
                  <div
                    style={
                      Object {
                        "alignItems": "center",
                        "background": "#e1e1e1",
                        "borderRadius": "4px",
                        "display": "flex",
                        "height": "200px",
                        "justifyContent": "center",
                        "width": "260px",
                      }
                    }
                  >
                    <img
                      alt="Place image"
                      className="place-card__image"
                      height={0}
                      onError={[Function]}
                      onLoad={[Function]}
                      src="img/apartment-01.jpg"
                      width={0}
                    />
                    <span>
                      Loading ...
                    </span>
                  </div>
                </a>
              </div>
              <div
                className="place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      120
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button"
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "92.00%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="#"
                    onClick={[Function]}
                  >
                    The Pondhouse - A Magical Place
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  house
                </p>
              </div>
            </article>
          </div>
        </section>
        <div
          className="cities__right-section"
        >
          <section
            className="cities__map map"
          >
            <div
              id="map"
              style={
                Object {
                  "height": "100%",
                  "top": 0,
                  "zIndex": 100,
                }
              }
            />
          </section>
        </div>
      </div>
    </div>
  </main>
</div>
`;
